test/es6/iteratorclose.js

```diff
@@ -0,0 +1,950 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+WScript.LoadScriptFile("..\\UnitTestFramework\\UnitTestFramework.js");
+
+var nextCount = 0;
+var returnCount = 0;
+var value1 = 1;
+var done1 = false;
+var iterable = {};
+var shouldNextThrow = false;
+var shouldReturnThrow = false;
+iterable[Symbol.iterator] = function () {
+    return {
+        next: function() {
+            nextCount++;
+            if (shouldNextThrow) { throw new Error('Exception from next function'); }
+            return {value : value1, done:done1};
+        },
+        return: function (value) {
+            returnCount++;
+            if (shouldReturnThrow) { throw new Error('Exception from return function'); }
+            return {done:true};
+        }
+    };
+};
+
+var obj = {
+  set prop(val) {
+    throw new Error('From setter');
+  }
+};
+
+var tests = [
+    {
+		name : "Destructuring - no .return function defined is a valid scenario",
+		body : function () {
+			var iterable = {};
+			iterable[Symbol.iterator] = function () {
+				return {
+					next : function () {
+						return {
+							value : 10,
+							done : false
+						};
+					}
+				};
+			};
+
+			var [a] = iterable;
+			assert.areEqual(a, 10);
+            var b;
+            [b] = iterable;
+			assert.areEqual(b, 10);
+		}
+
+	},
+    {
+		name : "Destructuring - validating that the iterable has .return field but not as a function",
+		body : function () {
+            var returnFnc = undefined;
+            var val = 20;
+            var iterable = {};
+            iterable[Symbol.iterator] = function() {
+              return {
+                next: function() {
+                    return {value:val, done:false};
+                },
+                return: returnFnc
+              };
+            };
+            
+            var a;
+            [a] = iterable;
+            assert.areEqual(a, 20);
+            
+            returnFnc = null;

```

Comment:
This doesn't affect the object pointed to by `iterable` at all. You need to set `iterable.return` to null directly